### PR TITLE
Update dependencies

### PR DESCRIPTION
**Pull request by ralphtheninja**
Closes https://github.com/Level/subleveldown/issues/9
Closes https://github.com/Level/subleveldown/issues/24
Closes https://github.com/Level/subleveldown/issues/22
Supersedes https://github.com/Level/subleveldown/pull/7

Squash when merge.

### CODE REVIEW

**Comment by ralphtheninja** [comment]
What's left now are some minor cleanup, docs, changelog, upgrade guide etc.

**Comment by ralphtheninja** [comment]
And later on refactoring to get rid of `down()` function.

File: leveldown.js [hunk 1]
```diff
@@ -83,7 +83,8 @@ SubDown.prototype._open = function (opts, cb) {
     return done()
   }
 
-  this.db.on('open', this.open.bind(this, opts, done))
+  this.db.once('open', this.open.bind(this, opts, done))
+  this.db.open()
```
Comment (ralphtheninja) [inline leveldown.js]: This fixes the open, close, open, close test in `abstract-leveldown` (which @finnp commented out in previous PR)

Comment (ralphtheninja) [reply leveldown.js]: Maybe we should tweak it to do `if (this.db.isClosed()) { this.db.open() }`?

**Comment by vweevers** [comment]
I'll review tomorrow.

File: test/index.js [hunk 2]
```diff
@@ -65,25 +65,26 @@ test('SubDown constructor', function (t) {
 test('SubDb main function', function (t) {
   t.test('opts.open hook', function (t) {
     t.plan(1)
-    subdb(levelup('loc', {db: memdown}), 'test', {
+    subdb(levelup(memdown('loc')), 'test', {
```
Comment (vweevers) [inline test/index.js]: Can remove the location argument (5x)

File: leveldown.js (hunk 1 above)
Comment (vweevers) [reply leveldown.js]: @ralphtheninja would this work?

```js
SubDown.prototype._open = function (opts, cb) {
  var self = this

  this.db.open(function (err) {
    if (err) return cb(err)

    var subdb = down(self.db, 'subleveldown')

    if (subdb && subdb.prefix) {
      self.prefix = subdb.prefix + self.prefix
      self.leveldown = down(subdb.db)
    } else {
      self.leveldown = down(self.db)
    }

    if (self._beforeOpen) self._beforeOpen(cb)
    else cb()
  })
}
```

Comment (vweevers) [reply leveldown.js]: @ralphtheninja ping in case you missed this because GH hides it

Comment (ralphtheninja) [reply leveldown.js]: Much cleaner! Let me try it out.

Comment (ralphtheninja) [reply leveldown.js]: Updated.

**Comment by vweevers** [comment]
Could `_serializeKey` and `_serializeValue` be(come) a problem?

**Comment by ralphtheninja** [comment]
@vweevers You mean the default in `abstract-leveldown`?

```js
AbstractLevelDOWN.prototype._serializeKey = function (key) {              
  return Buffer.isBuffer(key) ? key : String(key)                         
}                                                                         
                                                                          
AbstractLevelDOWN.prototype._serializeValue = function (value) {          
  if (value == null) return ''                                            
  return Buffer.isBuffer(value) || process.browser ? value : String(value)
}                                                                         
```

**Comment by ralphtheninja** [comment]
`_serializeKey` looks ok for Buffer case but might be a problem with keys that aren't strings, OTOH `concat` doesn't care about `prefix` if it's not a string or a Buffer:

```js
var concat = function (prefix, key, force) {                               
  if (typeof key === 'string' && (force || key.length)) return prefix + key
  if (Buffer.isBuffer(key) && (force || key.length)) {                     
    return Buffer.concat([Buffer.from(prefix), key])                       
  }                                                                        
  return key                                                               
}                                                                          
```

**Comment by ralphtheninja** [comment]
@vweevers Also, remember that `levelup` + `encoding-down` is in front of all these `._put()`, `._get()` operations. So as long as the encoding from `levelup` -> `subdown` makes keys as strings or buffers, we should be good.

**Pull request merged**